File: tests/test_scenarios/rust/src/cit/snapshots.rs
use crate::helpers::kvs_hash_paths;
use crate::helpers::kvs_instance::kvs_instance;
use crate::helpers::kvs_parameters::KvsParameters;
use rust_kvs::prelude::{KvsApi, SnapshotId};
use serde_json::Value;
use test_scenarios_rust::scenario::{Scenario, ScenarioGroup, ScenarioGroupImpl};
use tracing::info;

struct SnapshotCount;

impl Scenario for SnapshotCount {
    fn name(&self) -> &str {
        "count"
    }

    fn run(&self, input: &str) -> Result<(), String> {
        let v: Value = serde_json::from_str(input).expect("Failed to parse input string");
        let count =
            serde_json::from_value(v["count"].clone()).expect("Failed to parse \"count\" field");
        let params = KvsParameters::from_value(&v).expect("Failed to parse parameters");

        // Create snapshots.
        for i in 0..count {
            let kvs = kvs_instance(params.clone()).expect("Failed to create KVS instance");
            kvs.set_value("counter", i).expect("Failed to set value");
            info!(snapshot_count = kvs.snapshot_count());

            // Flush KVS.
            kvs.flush().expect("Failed to flush");
        }

        {
            let kvs = kvs_instance(params).expect("Failed to create KVS instance");
            info!(snapshot_count = kvs.snapshot_count());
        }

        Ok(())
    }
}

struct SnapshotMaxCount;

impl Scenario for SnapshotMaxCount {
    fn name(&self) -> &str {
        "max_count"
    }

    fn run(&self, input: &str) -> Result<(), String> {
        let v: Value = serde_json::from_str(input).expect("Failed to parse input string");
        let params = KvsParameters::from_value(&v).expect("Failed to parse parameters");

        let kvs = kvs_instance(params.clone()).expect("Failed to create KVS instance");
        info!(max_count = kvs.snapshot_max_count());
        Ok(())
    }
}

struct SnapshotRestore;

impl Scenario for SnapshotRestore {
    fn name(&self) -> &str {
        "restore"
    }

    fn run(&self, input: &str) -> Result<(), String> {
        let v: Value = serde_json::from_str(input).expect("Failed to parse input string");
        let count =
            serde_json::from_value(v["count"].clone()).expect("Failed to parse \"count\" field");
        let snapshot_id = serde_json::from_value(v["snapshot_id"].clone())
            .expect("Failed to parse \"snapshot_id\" field");
        let params = KvsParameters::from_value(&v).expect("Failed to parse parameters");

        // Create snapshots.
        for i in 0..count {
            let kvs = kvs_instance(params.clone()).expect("Failed to create KVS instance");
            kvs.set_value("counter", i).expect("Failed to set value");

            // Flush KVS.
            kvs.flush().expect("Failed to flush");
        }

        {
            let kvs = kvs_instance(params).expect("Failed to create KVS instance");

            let result = kvs.snapshot_restore(SnapshotId(snapshot_id));
            info!(result = format!("{result:?}"));
            if let Ok(()) = result {
                let value = kvs
                    .get_value_as::<i32>("counter")
                    .expect("Failed to read value");
                info!(value);
            }
        }

        Ok(())
    }
}

struct SnapshotPaths;

impl Scenario for SnapshotPaths {
    fn name(&self) -> &str {
        "paths"
    }

    fn run(&self, input: &str) -> Result<(), String> {
        let v: Value = serde_json::from_str(input).expect("Failed to parse input string");
        let count =
            serde_json::from_value(v["count"].clone()).expect("Failed to parse \"count\" field");
        let snapshot_id = serde_json::from_value(v["snapshot_id"].clone())
            .expect("Failed to parse \"snapshot_id\" field");
        let params = KvsParameters::from_value(&v).expect("Failed to parse parameters");
        let instance_id = params.instance_id;
        let working_dir = params.dir.clone().expect("Working directory must be set");

        // Create snapshots.
        for i in 0..count {
            let kvs = kvs_instance(params.clone()).expect("Failed to create KVS instance");
            kvs.set_value("counter", i).expect("Failed to set value");

            // Flush KVS.
            kvs.flush().expect("Failed to flush");
        }

        {
            let (kvs_path, hash_path) =
                kvs_hash_paths(&working_dir, instance_id, SnapshotId(snapshot_id));
            info!(
                kvs_path = format!("{}", kvs_path.display()),
                hash_path = format!("{}", hash_path.display())
            );
        }

        Ok(())
    }
}

struct SnapshotIDAssignment;

impl Scenario for SnapshotIDAssignment {
    fn name(&self) -> &str {
        "id_assignment"
    }

    fn run(&self, input: &str) -> Result<(), String> {
        let v: Value = serde_json::from_str(input).expect("Failed to parse input string");
        let count: i32 =
            serde_json::from_value(v["count"].clone()).expect("Failed to parse \"count\" field");
        let params = KvsParameters::from_value(&v).expect("Failed to parse parameters");

        // Create snapshots and track their IDs.
        for i in 0..count {
            let kvs = kvs_instance(params.clone()).expect("Failed to create KVS instance");
            kvs.set_value("counter", i).expect("Failed to set value");
            kvs.flush().expect("Failed to flush");
        }

        // Verify snapshot IDs exist
        let kvs = kvs_instance(params.clone()).expect("Failed to create KVS instance");
        let snapshot_count = kvs.snapshot_count();
        info!(snapshot_ids = format!("count={}", snapshot_count));

        Ok(())
    }
}

struct SnapshotDeletion;

impl Scenario for SnapshotDeletion {
    fn name(&self) -> &str {
        "deletion"
    }

    fn run(&self, input: &str) -> Result<(), String> {
        let v: Value = serde_json::from_str(input).expect("Failed to parse input string");
        let count: i32 =
            serde_json::from_value(v["count"].clone()).expect("Failed to parse \"count\" field");
        let params = KvsParameters::from_value(&v).expect("Failed to parse parameters");
        let snapshot_max_count = kvs_instance(params.clone())
            .expect("Failed to create KVS instance")
            .snapshot_max_count();

        // Create more snapshots than the maximum to trigger deletion.
        for i in 0..count {
            let kvs = kvs_instance(params.clone()).expect("Failed to create KVS instance");
            kvs.set_value("counter", i).expect("Failed to set value");
            kvs.flush().expect("Failed to flush");
        }

        // Verify that only max_count snapshots exist
        let kvs = kvs_instance(params).expect("Failed to create KVS instance");
        let final_snapshot_count = kvs.snapshot_count();

        info!(
            oldest_deleted =
                final_snapshot_count == snapshot_max_count && count > snapshot_max_count as i32
        );

        Ok(())
    }
}

struct SnapshotNoVersioning;

impl Scenario for SnapshotNoVersioning {
    fn name(&self) -> &str {
        "no_versioning"
    }

    fn run(&self, input: &str) -> Result<(), String> {
        let v: Value = serde_json::from_str(input).expect("Failed to parse input string");
        let params = KvsParameters::from_value(&v).expect("Failed to parse parameters");

        // Create a KVS and flush it
        let kvs = kvs_instance(params.clone()).expect("Failed to create KVS instance");
        kvs.set_value("test_key", 42).expect("Failed to set value");
        kvs.flush().expect("Failed to flush");

        // Read the JSON file and verify no version field exists
        let instance_id = params.instance_id;
        let working_dir = params.dir.expect("Working directory must be set");
        let (kvs_path, _) = kvs_hash_paths(&working_dir, instance_id, SnapshotId(0));

        let file_content = std::fs::read_to_string(&kvs_path).expect("Failed to read KVS file");
        let kvs_data: Value =
            serde_json::from_str(&file_content).expect("Failed to parse KVS JSON");

        // Check that there's no 'version' field
        let no_version_field = !kvs_data
            .as_object()
            .is_some_and(|obj| obj.contains_key("version"));
        info!(no_version_field);

        Ok(())
    }
}

pub fn snapshots_group() -> Box<dyn ScenarioGroup> {
    Box::new(ScenarioGroupImpl::new(
        "snapshots",
        vec![
            Box::new(SnapshotCount),
            Box::new(SnapshotMaxCount),
            Box::new(SnapshotRestore),
            Box::new(SnapshotPaths),
            Box::new(SnapshotIDAssignment),
            Box::new(SnapshotDeletion),
            Box::new(SnapshotNoVersioning),
        ],
        vec![],
    ))
}

File: tests/test_scenarios/cpp/src/cit/snapshots.cpp
/********************************************************************************
 * Copyright (c) 2025 Contributors to the Eclipse Foundation
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information regarding copyright ownership.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: Apache-2.0
 ********************************************************************************/

#include "snapshots.hpp"
#include "helpers/helpers.hpp"
#include "helpers/kvs_instance.hpp"
#include "helpers/kvs_parameters.hpp"
#include "tracing.hpp"

using namespace score::mw::per::kvs;
using namespace score::json;

namespace
{
const std::string kTargetName{"cpp_test_scenarios::snapshots::count"};

template <typename T>
T get_field(const Object& obj, const std::string& field)
{
    auto it{obj.find(field)};
    if (it == obj.end())
    {
        throw std::runtime_error("Missing field: " + field);
    }
    return it->second.As<T>().value();
}

Object get_object(const std::string& data)
{
    JsonParser parser;
    auto from_buffer_result{parser.FromBuffer(data)};
    if (!from_buffer_result)
    {
        throw std::runtime_error{"Failed to parse JSON"};
    }

    auto as_object_result{from_buffer_result.value().As<Object>()};
    if (!as_object_result)
    {
        throw std::runtime_error{"Failed to cast JSON to object"};
    }

    return std::move(as_object_result.value().get());
}
}  // namespace

class SnapshotCount : public Scenario
{
  public:
    ~SnapshotCount() final = default;

    std::string name() const final
    {
        return "count";
    }

    void run(const std::string& input) const final
    {
        auto obj{get_object(input)};
        auto count{get_field<int32_t>(obj, "count")};
        auto params{KvsParameters::from_json(input)};

        // Create snapshots.
        for (int32_t i{0}; i < count; ++i)
        {
            auto kvs{kvs_instance(params)};
            auto set_result{kvs.set_value("counter", KvsValue{i})};
            if (!set_result)
            {
                throw std::runtime_error{"Failed to set value"};
            }

            auto count_result{kvs.snapshot_count()};
            if (!count_result)
            {
                throw std::runtime_error{"Unable to get snapshot count"};
            }

            TRACING_INFO(kTargetName, std::pair{std::string{"snapshot_count"}, count_result.value()});

            // Flush KVS.
            auto flush_result{kvs.flush()};
            if (!flush_result)
            {
                throw std::runtime_error{"Failed to flush"};
            }

            // Create snapshot in the first available slot. Ignore failure if max count is reached.
            kvs.snapshot_create();
        }

        {
            auto kvs{kvs_instance(params)};
            auto count_result{kvs.snapshot_count()};
            if (!count_result)
            {
                throw std::runtime_error{"Unable to get snapshot count"};
            }
            TRACING_INFO(kTargetName, std::pair{std::string{"snapshot_count"}, count_result.value()});
        }
    }
};

class SnapshotMaxCount : public Scenario
{
  public:
    ~SnapshotMaxCount() final = default;

    std::string name() const final
    {
        return "max_count";
    }

    void run(const std::string& input) const final
    {
        auto params{KvsParameters::from_json(input)};

        auto kvs{kvs_instance(params)};
        TRACING_INFO(kTargetName, std::pair{std::string{"max_count"}, kvs.snapshot_max_count()});
    }
};

class SnapshotRestore : public Scenario
{
  public:
    ~SnapshotRestore() final = default;

    std::string name() const final
    {
        return "restore";
    }

    void run(const std::string& input) const final
    {
        auto obj{get_object(input)};
        auto count{get_field<int32_t>(obj, "count")};
        auto snapshot_id{get_field<uint64_t>(obj, "snapshot_id")};
        auto params{KvsParameters::from_json(input)};

        // Create snapshots.
        for (int32_t i{0}; i < count; ++i)
        {
            auto kvs{kvs_instance(params)};
            auto set_result{kvs.set_value("counter", KvsValue{i})};
            if (!set_result)
            {
                throw std::runtime_error{"Failed to set value"};
            }

            // Flush KVS.
            auto flush_result{kvs.flush()};
            if (!flush_result)
            {
                throw std::runtime_error{"Failed to flush"};
            }

            // Create snapshot in the first available slot.
            auto create_result{kvs.snapshot_create()};
            if (!create_result)
            {
                throw std::runtime_error{"Failed to create snapshot"};
            }
        }

        {
            auto kvs{kvs_instance(params)};

            auto restore_result{kvs.snapshot_restore(snapshot_id)};
            TRACING_INFO(kTargetName,
                         std::pair{std::string{"result"}, restore_result ? "Ok(())" : "Err(InvalidSnapshotId)"});

            if (restore_result)
            {
                auto get_result{kvs.get_value("counter")};
                if (!get_result)
                {
                    throw std::runtime_error{"Failed to read value"};
                }

                auto value{std::get<int32_t>(get_result.value().getValue())};
                TRACING_INFO(kTargetName, std::pair{std::string{"value"}, value});
            }
        }
    }
};

class SnapshotPaths : public Scenario
{
  public:
    ~SnapshotPaths() final = default;

    std::string name() const final
    {
        return "paths";
    }

    void run(const std::string& input) const final
    {
        auto obj{get_object(input)};
        auto count{get_field<int32_t>(obj, "count")};
        auto snapshot_id{get_field<uint64_t>(obj, "snapshot_id")};
        auto params{KvsParameters::from_json(input)};
        auto working_dir{*params.dir};
        auto instance_id{params.instance_id};

        // Create snapshots.
        for (int32_t i{0}; i < count; ++i)
        {
            auto kvs{kvs_instance(params)};
            auto set_result{kvs.set_value("counter", KvsValue{i})};
            if (!set_result)
            {
                throw std::runtime_error{"Failed to set value"};
            }

            // Flush KVS.
            auto flush_result{kvs.flush()};
            if (!flush_result)
            {
                throw std::runtime_error{"Failed to flush"};
            }

            // Create snapshot in the first available slot.
            auto create_result{kvs.snapshot_create()};
            if (!create_result)
            {
                throw std::runtime_error{"Failed to create snapshot"};
            }
        }

        {
            auto [kvs_path, hash_path] = kvs_hash_paths(working_dir, instance_id, snapshot_id);
            TRACING_INFO(kTargetName,
                         std::make_pair(std::string{"kvs_path"}, kvs_path),
                         std::make_pair(std::string{"hash_path"}, hash_path));
        }
    }
};

class SnapshotCreate : public Scenario
{
  public:
    ~SnapshotCreate() final = default;

    std::string name() const final
    {
        return "create";
    }

    void run(const std::string& input) const final
    {
        auto params{KvsParameters::from_json(input)};

        auto kvs{kvs_instance(params)};
        auto set_result{kvs.set_value("counter", KvsValue{42})};
        if (!set_result)
        {
            throw std::runtime_error{"Failed to set value"};
        }

        // Flush only: in C++ this does NOT create a snapshot.
        auto flush_result{kvs.flush()};
        if (!flush_result)
        {
            throw std::runtime_error{"Failed to flush"};
        }

        auto count_after_flush{kvs.snapshot_count()};
        if (!count_after_flush)
        {
            throw std::runtime_error{"Unable to get snapshot count"};
        }
        TRACING_INFO(kTargetName,
                     std::pair{std::string{"snapshot_count_after_flush"}, count_after_flush.value()});

        // Explicitly create snapshot.
        auto create_result{kvs.snapshot_create()};
        TRACING_INFO(kTargetName,
                     std::pair{std::string{"result"}, create_result ? "Ok(())" : "Err(OutOfStorageSpace)"});

        auto count_after_create{kvs.snapshot_count()};
        if (!count_after_create)
        {
            throw std::runtime_error{"Unable to get snapshot count"};
        }
        TRACING_INFO(kTargetName,
                     std::pair{std::string{"snapshot_count_after_create"}, count_after_create.value()});
    }
};

class SnapshotDelete : public Scenario
{
  public:
    ~SnapshotDelete() final = default;

    std::string name() const final
    {
        return "delete";
    }

    void run(const std::string& input) const final
    {
        auto obj{get_object(input)};
        auto count{get_field<int32_t>(obj, "count")};
        auto snapshot_id{get_field<uint64_t>(obj, "snapshot_id")};
        auto params{KvsParameters::from_json(input)};

        // Create snapshots.
        for (int32_t i{0}; i < count; ++i)
        {
            auto kvs{kvs_instance(params)};
            auto set_result{kvs.set_value("counter", KvsValue{i})};
            if (!set_result)
            {
                throw std::runtime_error{"Failed to set value"};
            }

            // Flush KVS.
            auto flush_result{kvs.flush()};
            if (!flush_result)
            {
                throw std::runtime_error{"Failed to flush"};
            }

            // Create snapshot in the first available slot.
            auto create_result{kvs.snapshot_create()};
            if (!create_result)
            {
                throw std::runtime_error{"Failed to create snapshot"};
            }
        }

        {
            auto kvs{kvs_instance(params)};

            auto delete_result{kvs.snapshot_delete(snapshot_id)};
            TRACING_INFO(kTargetName,
                         std::pair{std::string{"result"}, delete_result ? "Ok(())" : "Err(InvalidSnapshotId)"});

            if (delete_result)
            {
                auto count_result{kvs.snapshot_count()};
                if (!count_result)
                {
                    throw std::runtime_error{"Unable to get snapshot count after delete"};
                }
                TRACING_INFO(kTargetName, std::pair{std::string{"snapshot_count"}, count_result.value()});
            }
        }
    }
};

ScenarioGroup::Ptr snapshots_group()
{
    return ScenarioGroup::Ptr{new ScenarioGroupImpl{"snapshots",
                                                    {std::make_shared<SnapshotCount>(),
                                                     std::make_shared<SnapshotMaxCount>(),
                                                     std::make_shared<SnapshotRestore>(),
                                                     std::make_shared<SnapshotPaths>(),
                                                     std::make_shared<SnapshotCreate>(),
                                                     std::make_shared<SnapshotDelete>()},
                                                    {}}};
}
